### PR TITLE
[MAINTENANCE] Add `get_all` method to Store and StoreBackend APIs

### PR DESCRIPTION
**Pull request by cdkini**
Retrieval of an entire collection is not currently supported by our stores - we should allow for a `get_all` mechanism to improve performance (especially around datasources).

- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [x] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: great_expectations/data_context/store/_store_backend.py [hunk 1]
```diff
@@ -216,6 +221,10 @@ def _validate_value(self, value) -> None:  # noqa: B027 # no abstract decorator
     def _get(self, key) -> None:
         raise NotImplementedError
 
+    @abstractmethod
+    def _get_all(self) -> list[Any]:
+        raise NotImplementedError
```
Comment (Kilo59) [inline great_expectations/data_context/store/_store_backend.py:226]: `#future-enhancement`

Can we make the StoreBackend Generic and then we don't have to use `Any` for the return types?

But this should be done in a different PR. I don't want to slow down this PR from getting merged.

https://docs.python.org/3/library/typing.html#user-defined-generic-types

```python
from typing import TypeVar, Generic

T = TypeVar("T")

class StoreBackend(Generic[T], metaclass=ABCMeta):

    def get_all(self) -> T:
        return self._get_all()

    @abstractmethod
    def _get_all(self) -> list[T]:
        raise NotImplementedError

````

File: tests/data_context/store/test_gx_cloud_store_backend.py [hunk 2]
```diff
@@ -271,7 +271,24 @@ def test_list_keys(
...
+            params=None,
+        )
+
+
+def test_get_all(
+    construct_ge_cloud_store_backend: Callable[
+        [GXCloudRESTResource], GXCloudStoreBackend
+    ],
+) -> None:
+    store_backend = construct_ge_cloud_store_backend(GXCloudRESTResource.DATASOURCE)
```
Comment (Kilo59) [inline tests/data_context/store/test_gx_cloud_store_backend.py:284]: `non-blocking`

Can you use a `responses` mock here?

https://github.com/getsentry/responses#shortcuts

**Review by Kilo59 — APPROVED** [review]
Thanks for this!

File: great_expectations/data_context/store/_store_backend.py (hunk 1 above)
Comment (cdkini) [reply great_expectations/data_context/store/_store_backend.py:226]: Hmm I agree with this but we're not actually returning hydrated objects (just dict configs) - we should be returning rich config types or actual datasources but our stores use primitives

File: tests/data_context/store/test_gx_cloud_store_backend.py (hunk 2 above)
Comment (cdkini) [reply tests/data_context/store/test_gx_cloud_store_backend.py:284]: I'll add this in the follow-up PR - thanks!

**Pull request merged**